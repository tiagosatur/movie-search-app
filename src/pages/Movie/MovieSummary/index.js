export { default as MovieSummary } from './MovieSummary';