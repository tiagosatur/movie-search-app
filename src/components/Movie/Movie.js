import React, { useEffect, useState } from 'react';
import { useSelector } from 'react-redux';
import { Link, Route } from 'react-router-dom';
import styled from 'styled-components';

import useActions from '../../utils/hooks/useActions';
import { LoadingSpinner } from '../LoadingSpinner'
import { DefaultInput } from '../Input'
import { Error } from '../Error'
import useFormInput from '../../utils/hooks/useFormInput';
import { MovieDetail } from './MovieDetail'
import { MovieSummary } from './MovieSummary'
import { Button } from '../Button';
import { Modal } from '../Modal';
import { rem } from '../../style';

const Movie = ({ match }) => {
    const [ showModal, setShowModal ] = useState(false);
    const { searchMovieAction } = useActions();
    const { values, handleInputChange } = useFormInput({
      searchValue: '',
    });
    
    const {
        movie: {
          movie,
          totalResults,
          pending,
          error,
        },
    } = useSelector(state => state);
    
    const handleSubmit = (e) => {
      e.preventDefault()
      const { searchValue } = values;

      if(searchValue) {
        searchMovieAction(searchValue);
      }
     
    }

    const toggleModal = () => {
      setShowModal(!showModal);
    }
    /* if(error) {
      searchMovieClearAction();
    } */

    return (
        <>
            <div>
              <StyledForm onSubmit={handleSubmit}>
                <DefaultInput 
                    name='searchValue' 
                    value={values.searchValue}
                    placeholder='Search a movie...'
                    handleChange={handleInputChange} 
                  />
                <Button solid rounded right >Search</Button>
              </StyledForm>
            </div>

            { pending && <LoadingSpinner medium />}

            { error && <Error message={error} /> }

            { movie && movie.length > 0 && !pending && !error && (
              <p>
                We found <strong>{ totalResults }</strong>
                {totalResults == 1 ? 'result!' : ' results!'}
              </p>
            )}

            <StyledMovies>
            
              {movie && movie.map((m) => {
                  const { Title, Poster, Plot, imdbID } = m

                  return(
                    <StyledMovieItem key={Math.floor(Math.random())}>
                      <Link to={`${match.url}/${imdbID}`} key={Math.floor(Math.random())} onClick={setShowModal}>
                        <MovieSummary data={{Title, Poster, Plot}} />
                      </Link>
                      <Modal handleClose={toggleModal} show={showModal}>
                        <Route
                          exact
                          path={`${ match.path }/:imdbID`}
                          render={(props) => <MovieDetail data={movie} {...props} /> }
                        />
                      </Modal>
                    </StyledMovieItem>
                  )
                })
              }
            </StyledMovies>
        </>
    )
    
}

const StyledMovies = styled.div`
  display: flex;
  flex-wrap: wrap;
`;
const StyledMovieItem = styled.div`
  margin: ${rem(16)};
  width: 20%;
`;

const StyledForm = styled.form`
  display: flex;
  justify-content: space-between;
  margin: ${rem(20)} auto 0 auto;
  width: 100%;

  @media(min-width: ${rem(768)}) {
    margin: ${rem(100)} auto 0 auto;
    width: 50%;
  }

  .input {
    flex-grow: 1;
  }
  .button {

  }
`;


export default Movie;