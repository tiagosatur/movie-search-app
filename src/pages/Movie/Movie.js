import React, { useState } from 'react';
import { useSelector, connect } from 'react-redux';
import { Link, Route } from 'react-router-dom';
import styled from 'styled-components';
import { withRouter } from "react-router";
import uniqid from 'uniqid';

import { searchMovie } from '../../redux/actions'

import useActions from '../../utils/hooks/useActions';
import { 
  Button, DefaultInput, Modal, Error, LoadingSpinner 
} from '../../components'
import useFormInput from '../../utils/hooks/useFormInput';
import { MovieDetail } from './MovieDetail'
import { MovieSummary } from './MovieSummary'
import { rem } from '../../style';


export const Movie = ({
   match, 
   searchMovieAction,
   movies,
   totalResults,
   pending,
   error,
  }, props) => {
    const [ showModal, setShowModal ] = useState(false);
    //const { searchMovieAction } = useActions();
    const { values, handleInputChange } = useFormInput({
      searchValue: '', 
    });
    
    /* const {
        movie: {
          movie,
          totalResults,
          pending,
          error,
        },
    } = useSelector(state => state); */
    
    const handleSubmit = (e) => {
      e.preventDefault()
      const { searchValue } = values;

      if(searchValue) {
        searchMovieAction(searchValue);
      }
    }

    const toggleModal = () => {
      setShowModal(!showModal);
    }

    return (
        <div className='movies'>
            <div>
              <StyledForm className='search-form' onSubmit={handleSubmit}>
                <DefaultInput 
                    name='searchValue' 
                    value={values.searchValue}
                    placeholder='Search a movie...'
                    handleChange={handleInputChange} 
                  />
                <Button solid rounded right >Search</Button>
              </StyledForm>
            </div>

            { pending && <LoadingSpinner medium />}

            { error && <Error message={error} /> }

            { movies && movies.length > 0 && !pending && !error && (
              <p>
                We found <strong>{ totalResults }</strong>
                {totalResults == 1 ? 'result!' : ' results!'}
              </p>
            )}

            <StyledMovies>
            
              {movies && movies.map((m) => {
                  const { Title, Poster, Plot, imdbID } = m

                  return(
                    <StyledMovieItem key={uniqid()} className='search-list__item'>
                      <Link to={`${match.url}/${imdbID}`} onClick={setShowModal}>
                        <MovieSummary data={{Title, Poster, Plot}} />
                      </Link>
                      
                    </StyledMovieItem>
                  )
                })
              }
              <Modal handleClose={toggleModal} show={showModal}>
                <Route
                  exact
                  path={`${ match.path }/:imdbID`}
                  render={(props) => <MovieDetail {...props} /> }
                />
              </Modal>
            </StyledMovies>
        </div>
    )
    
}

const StyledMovies = styled.div`
  display: flex;
  flex-wrap: wrap;
`;
export const StyledMovieItem = styled.div`
  margin: ${rem(16)};
  width: 20%;
`;

const StyledForm = styled.form`
  display: flex;
  justify-content: space-between;
  margin: ${rem(20)} auto 0 auto;
  width: 100%;

  @media(min-width: ${rem(768)}) {
    margin: ${rem(100)} auto 0 auto;
    width: 50%;
  }

  .input {
    flex-grow: 1;
  }
  .button {

  }
`;

const mapStateToProps = state => {
  return {
    movies: state.movie.list,
    totalResults: state.movie.totalResults,
    pending: state.movie.pending,
    error: state.movie.error,
  }
};

const mapDispatchToProps = dispatch => ({
  searchMovieAction: data => dispatch(searchMovie(data))
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withRouter(Movie));