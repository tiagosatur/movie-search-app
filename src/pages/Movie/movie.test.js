import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Movie from './Movie';

describe('Movie Page tests', () => {
    let wrapper;
    const props = {};
    beforeEach(() => {
        wrapper = shallow(
            <Movie /> 
        ).dive();
    });

    it('should render the Movie Page', () => {
        expect(wrapper.find('.Movie').exists()).toBe(true);
    });

    it('should call Movie snapshot correctly', () => {
        const tree = renderer
            .create(wrapper)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
})