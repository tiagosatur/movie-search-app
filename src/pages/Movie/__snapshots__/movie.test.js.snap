// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie Page tests should call Movie snapshot correctly 1`] = `null`;
